File: sdk/agentlens/cost.py
"""
Cost estimation.

The important behaviour here isn't the prices — those go stale the moment
they're written. It's the distinction between **zero** and **unknown**.

An unrecognized model used to estimate at $0.00, which reads as "this step
was free" when it means "nobody priced this." A dashboard summing those
zeros reports a total that's confidently wrong, and the error is invisible
because $0.00 is a perfectly plausible number. So every estimate carries
where it came from, and anything unpriced is surfaced rather than absorbed.

Prices are USD per 1M tokens (input, output). Override per-client with
`AgentLens(cost_table={...})`, or globally with `AGENTLENS_COST_TABLE`
pointing at a JSON file — because a table baked into a release is a table
that's wrong by the next repricing.
"""

from __future__ import annotations

import json
import os
from typing import Optional

CostTable = "dict[str, tuple[float, float]]"

DEFAULT_COST_TABLE: dict[str, tuple[float, float]] = {
    # OpenAI
    "gpt-4o": (2.50, 10.00),
    "gpt-4o-mini": (0.15, 0.60),
    "gpt-4.1": (2.00, 8.00),
    "gpt-4.1-mini": (0.40, 1.60),
    "o3-mini": (1.10, 4.40),
    # Anthropic
    "claude-3-5-haiku": (0.80, 4.00),
    "claude-sonnet-4": (3.00, 15.00),
    "claude-opus-4": (15.00, 75.00),
    # Meta / open weights (typical hosted pricing)
    "llama-3.1-70b": (0.60, 0.80),
    "llama-3.1-8b": (0.10, 0.20),
    "mixtral-8x7b": (0.50, 0.50),
}

# Where a cost figure came from. Stored on the span so a total can say how
# much of itself it actually knows.
COST_REPORTED = "reported"  # the provider told us; trust it over any table
COST_TABLE = "table"  # matched a price entry
COST_UNPRICED = "unpriced"  # no entry — the 0.0 here is a gap, not a fact
COST_FREE = "free"  # priced at zero on purpose (local models)


def load_cost_table_from_env(env: Optional[dict] = None) -> dict:
    """
    Read extra prices from `AGENTLENS_COST_TABLE`: a path to a JSON file, or
    inline JSON. Shaped `{"model-substring": [input_per_1m, output_per_1m]}`.

    A malformed table is ignored rather than raised: prices are an
    enhancement, and failing an agent's startup over a typo in a pricing
    file would be a bad trade.
    """
    raw = (env if env is not None else os.environ).get("AGENTLENS_COST_TABLE", "").strip()
    if not raw:
        return {}

    try:
        if raw.startswith("{"):
            data = json.loads(raw)
        else:
            with open(raw, encoding="utf-8") as f:
                data = json.load(f)
        table = {}
        for model, prices in data.items():
            if isinstance(prices, (list, tuple)) and len(prices) == 2:
                table[str(model).lower()] = (float(prices[0]), float(prices[1]))
        return table
    except Exception:
        return {}


def lookup_price(model: str, cost_table: Optional[dict] = None) -> Optional[tuple]:
    """
    Find the price entry for a model, or None if nothing matches.

    Longest key first, so `gpt-4o-mini` doesn't get billed at `gpt-4o` rates
    — a 16x error that would look entirely plausible on a dashboard.
    """
    table = {**DEFAULT_COST_TABLE, **load_cost_table_from_env(), **(cost_table or {})}
    lowered = (model or "").lower()
    if not lowered:
        return None
    for key in sorted(table, key=len, reverse=True):
        if key in lowered:
            return table[key]
    return None


def estimate_cost(
    model: str,
    input_tokens: int,
    output_tokens: int,
    cost_table: Optional[dict] = None,
    reported_cost: Optional[float] = None,
) -> tuple[float, str]:
    """
    Return `(cost_usd, source)`.

    A provider-reported cost always wins: it's authoritative, and a local
    table is a guess about someone else's billing.
    """
    if reported_cost is not None:
        try:
            return round(float(reported_cost), 6), COST_REPORTED
        except (TypeError, ValueError):
            pass

    price = lookup_price(model, cost_table)
    if price is None:
        # 0.0 with an "unpriced" marker, so a caller can tell this apart
        # from a step that genuinely cost nothing
        return 0.0, COST_UNPRICED

    input_price, output_price = price
    if input_price == 0 and output_price == 0:
        return 0.0, COST_FREE
    return (input_tokens * input_price + output_tokens * output_price) / 1_000_000, COST_TABLE


def estimate_cost_usd(
    model: str,
    input_tokens: int,
    output_tokens: int,
    cost_table: Optional[dict] = None,
) -> float:
    """Cost only. Kept for callers that don't need the provenance."""
    return estimate_cost(model, input_tokens, output_tokens, cost_table)[0]


def extract_reported_cost(result: object) -> Optional[float]:
    """
    Pull a cost the provider already calculated, if the response carries one.

    Several gateways (OpenRouter, LiteLLM, Helicone) return the real charge,
    which beats any local estimate.
    """
    if not isinstance(result, dict):
        usage = getattr(result, "usage", None)
        for holder in (result, usage):
            for attr in ("cost", "total_cost", "cost_usd"):
                value = getattr(holder, attr, None)
                if isinstance(value, (int, float)):
                    return float(value)
        return None

    for key in ("cost", "total_cost", "cost_usd"):
        if isinstance(result.get(key), (int, float)):
            return float(result[key])
    usage = result.get("usage")
    if isinstance(usage, dict):
        for key in ("cost", "total_cost", "cost_usd"):
            if isinstance(usage.get(key), (int, float)):
                return float(usage[key])
    return None

File: sdk-ts/src/cost.ts
/** Best-effort cost estimation. Prices are USD per 1M tokens (input, output). */

export type CostTable = Record<string, [number, number]>;

export const DEFAULT_COST_TABLE: CostTable = {
  'gpt-4o': [2.5, 10.0],
  'gpt-4o-mini': [0.15, 0.6],
  'gpt-4.1': [2.0, 8.0],
  'gpt-4.1-mini': [0.4, 1.6],
  'o3-mini': [1.1, 4.4],
  'claude-3-5-haiku': [0.8, 4.0],
  'claude-sonnet-4': [3.0, 15.0],
  'claude-opus-4': [15.0, 75.0],
  'llama-3.1-70b': [0.6, 0.8],
  'llama-3.1-8b': [0.1, 0.2],
  'mixtral-8x7b': [0.5, 0.5],
};

/**
 * Where a cost figure came from.
 *
 * An unrecognized model estimates at 0, which reads as "this was free" when
 * it means "nobody priced this". Carrying the provenance lets a total say
 * how much of itself it actually knows.
 */
export type CostSource = 'reported' | 'table' | 'unpriced' | 'free';

export function lookupPrice(model: string, overrides: CostTable = {}): [number, number] | null {
  const table = { ...DEFAULT_COST_TABLE, ...overrides };
  const lowered = (model || '').toLowerCase();
  if (!lowered) return null;
  // longest key first, so gpt-4o-mini isn't billed at gpt-4o rates
  const key = Object.keys(table)
    .sort((a, b) => b.length - a.length)
    .find((k) => lowered.includes(k));
  return key ? table[key]! : null;
}

export function estimateCost(
  model: string,
  inputTokens: number,
  outputTokens: number,
  overrides: CostTable = {},
  reportedCost?: number | null,
): { cost: number; source: CostSource } {
  // a provider-reported cost is authoritative; a local table is a guess
  // about someone else's billing
  if (typeof reportedCost === 'number' && Number.isFinite(reportedCost)) {
    return { cost: Math.round(reportedCost * 1e6) / 1e6, source: 'reported' };
  }

  const price = lookupPrice(model, overrides);
  if (!price) return { cost: 0, source: 'unpriced' };

  const [inPrice, outPrice] = price;
  if (inPrice === 0 && outPrice === 0) return { cost: 0, source: 'free' };
  return { cost: (inputTokens * inPrice + outputTokens * outPrice) / 1_000_000, source: 'table' };
}

/** A cost the provider already calculated, if the response carries one. */
export function extractReportedCost(result: unknown): number | null {
  if (typeof result !== 'object' || result === null) return null;
  const r = result as Record<string, unknown>;
  const holders = [r, r['usage']];
  for (const holder of holders) {
    if (typeof holder !== 'object' || holder === null) continue;
    for (const key of ['cost', 'total_cost', 'cost_usd']) {
      const value = (holder as Record<string, unknown>)[key];
      if (typeof value === 'number' && Number.isFinite(value)) return value;
    }
  }
  return null;
}

export function estimateCostUsd(
  model: string,
  inputTokens: number,
  outputTokens: number,
  overrides: CostTable = {},
): number {
  return estimateCost(model, inputTokens, outputTokens, overrides).cost;
}

/** Pull token usage out of an OpenAI- or Anthropic-shaped response. */
export function extractUsage(result: unknown): { input: number; output: number; model: string } {
  const out = { input: 0, output: 0, model: '' };
  if (typeof result !== 'object' || result === null) return out;
  const r = result as Record<string, any>;
  if (typeof r['model'] === 'string') out.model = r['model'];
  const usage = r['usage'];
  if (typeof usage === 'object' && usage !== null) {
    const u = usage as Record<string, any>;
    out.input = Number(u['prompt_tokens'] ?? u['input_tokens'] ?? u['inputTokens'] ?? 0) || 0;
    out.output = Number(u['completion_tokens'] ?? u['output_tokens'] ?? u['outputTokens'] ?? 0) || 0;
  }
  return out;
}
